#  House Prices Dataset
---
**Name:** Ronald Wanjohi Gachoka <br>
**Student Number:** S2274825 <br>
**Course:** Software Development for Data Science <br>
**Assignment:** Coursework 1 <br>
**Program:** MSc Financial Technology 

## 1. Overview

This main objective of this coursework is to perform exploratory data analysis techniques on the 2 datasets that are related to residential properties.

The goal of this project will be:
> - To understand the factors that influence property price
> - Determine which year had the highest property 
     - Which month was the best to sell a property 
> - What type of properties were the best to own based on their sale price and neighborhood

**Methodology**
To achieve these goals, the coursework will involve the following steps:
> - Data understanding
> - Data pre-processing (cleaning)
> - Univariate and bivariate analysis to understand relationships and trends
> - Visualizations
> - Conclusion


## 2. Data Understanding

The 2 datasets presented to us are from the sale of properties in a small town in USA between 2006 and 2010.

The first dataset describes individual minute elements of the residential properties from:
- lot coverage, number of bedrooms, type of foundation, zoning class amongst many more

The 2nd dataset is much smaller but describes:
- the parcel number, type of property and the price it was sold at.

**Important notes** (from data dictionary)
- To deal with some of the clear outliers, we are recommended to remove any properties with more than 4,000 square feet.

**Load libraries**

In [1]:
# Data analysis
import pandas as pd
import numpy as np

# Statistics
import scipy.stats as stats

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Encoding
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

# Styling
sns.set_style("whitegrid")

In [2]:
# Importing data
housing_df = pd.read_csv("data/Housing_1.txt", sep="\t")
sales_df = pd.read_csv("data/Housing_2.txt", sep="\t")

# Replacing the space in columns names with "_"
housing_df.columns = housing_df.columns.str.replace(" ","_")
sales_df.columns = sales_df.columns.str.replace(" ","_")

# quick overview
housing_df.head()

Order        PID  MS_SubClass MS_Zoning  Lot_Frontage  Lot_Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot_Shape Land_Contour  ... Screen_Porch Pool_Area Pool_QC  Fence  \
0   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
1   NaN       Reg          Lvl  ...          120         0     NaN  MnPrv   
2   NaN       IR1          Lvl  ...            0         0     NaN    NaN   
3   NaN       Reg          Lvl  ...            0         0     NaN    NaN   
4   NaN       IR1          Lvl  ...            0         0     NaN  MnPrv   

  Misc_Feature Misc_Val Mo_Sold Yr_Sold  Sale_Type  Sale_Condition  
0          NaN        0       5    2010        WD           Normal  
1          NaN        0       6    2010        WD           Normal  
2         Gar2    12500       6    2010        WD           Normal  
3          NaN        0       4    2010        WD           Normal  
4          NaN        0       3    2010        WD           Normal  

[5 rows x 81 columns]

In [3]:
sales_df.head()

Order        PID  MS_SubClass  SalePrice
0      1  526301100           20     215000
1      2  526350040           20     105000
2      3  526351010           20     172000
3      4  526353030           20     244000
4      5  527105010           60     189900

Using `.head()` is useful as it helps quickly preview the data we have and check for any errors or unexpected values.

In the above case with housing_df we can see that there's a mixture of numerical and categorical data as well as some rows with a value of `0` which could be either missing values or a valid value.

In the case of sales_df, we can see that there's only numerical data and nothing out of the ordinary.

In [4]:
# Describe the data
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2933 entries, 0 to 2932
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2933 non-null   int64  
 1   PID              2933 non-null   int64  
 2   MS_SubClass      2933 non-null   int64  
 3   MS_Zoning        2933 non-null   object 
 4   Lot_Frontage     2443 non-null   float64
 5   Lot_Area         2933 non-null   int64  
 6   Street           2933 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot_Shape        2933 non-null   object 
 9   Land_Contour     2933 non-null   object 
 10  Utilities        2933 non-null   object 
 11  Lot_Config       2933 non-null   object 
 12  Land_Slope       2933 non-null   object 
 13  Neighborhood     2933 non-null   object 
 14  Condition_1      2933 non-null   object 
 15  Condition_2      2933 non-null   object 
 16  Bldg_Type        2933 non-null   object 
 17  House_Style   

In [5]:
# Describe the data
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2933 entries, 0 to 2932
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   Order        2933 non-null   int64
 1   PID          2933 non-null   int64
 2   MS_SubClass  2933 non-null   int64
 3   SalePrice    2933 non-null   int64
dtypes: int64(4)
memory usage: 91.8 KB


From the above lines of code, we see that:
- several columns have multiple missing values in the housing_df dataset
- some columns have the wrong data type
- there are no missing values in the sales_df dataset

In [6]:
# Shape of the data
print("Shape of the housing dataset:")
housing_df.shape

Shape of the housing dataset:


(2933, 81)

In [7]:
print("Shape of the sales_df dataset:")
sales_df.shape

Shape of the sales_df dataset:


(2933, 4)

From the above lines of code, we see that: 
- housing_df has 2933 rows and 81 columns
- sales_df has 2933 rows and 4 columns

In [8]:
# Summary statistics (numerical columns)
housing_df.describe()

Order           PID  MS_SubClass  Lot_Frontage       Lot_Area  \
count  2933.000000  2.933000e+03  2933.000000   2443.000000    2933.000000   
mean   1464.151381  7.142811e+08    57.349131     69.237822   10147.789976   
std     846.579702  1.887213e+08    42.632965     23.354888    7876.008408   
min       1.000000  5.263011e+08    20.000000     21.000000    1300.000000   
25%     731.000000  5.284770e+08    20.000000     58.000000    7441.000000   
50%    1464.000000  5.354532e+08    50.000000     68.000000    9439.000000   
75%    2197.000000  9.071811e+08    70.000000     80.000000   11553.000000   
max    2930.000000  1.007100e+09   190.000000    313.000000  215245.000000   

       Overall_Qual  Overall_Cond   Year_Built  Year_Remod/Add  Mas_Vnr_Area  \
count   2933.000000   2933.000000  2933.000000     2933.000000   2910.000000   
mean       6.094102      5.562564  1971.344016     1984.241050    102.082818   
std        1.410595      1.111114    30.232331       20.864849    179.172166   
min        1.000000      1.000000  1872.000000     1950.000000      0.000000   
25%        5.000000      5.000000  1954.000000     1965.000000      0.000000   
50%        6.000000      5.000000  1973.000000     1993.000000      0.000000   
75%        7.000000      6.000000  2001.000000     2004.000000    164.000000   
max       10.000000      9.000000  2010.000000     2010.000000   1600.000000   

       ...  Garage_Area  Wood_Deck_SF  Open_Porch_SF  Enclosed_Porch  \
count  ...  2932.000000   2933.000000    2933.000000     2933.000000   
mean   ...   472.749659     93.655984      47.501534       22.988067   
std    ...   214.957946    126.332461      67.457316       64.110459   
min    ...     0.000000      0.000000       0.000000        0.000000   
25%    ...   320.000000      0.000000       0.000000        0.000000   
50%    ...   480.000000      0.000000      27.000000        0.000000   
75%    ...   576.000000    168.000000      70.000000        0.000000   
max    ...  1488.000000   1424.000000     742.000000     1012.000000   

        3Ssn_Porch  Screen_Porch    Pool_Area      Misc_Val      Mo_Sold  \
count  2933.000000   2933.000000  2933.000000   2933.000000  2933.000000   
mean      2.589840     16.088987     2.241050     50.583362     6.270031   
std      25.128602     56.171408    35.579037    566.056789     3.570864   
min       0.000000      0.000000     0.000000      0.000000     1.000000   
25%       0.000000      0.000000     0.000000      0.000000     4.000000   
50%       0.000000      0.000000     0.000000      0.000000     6.000000   
75%       0.000000      0.000000     0.000000      0.000000     8.000000   
max     508.000000    576.000000   800.000000  17000.000000   120.000000   

           Yr_Sold  
count  2933.000000  
mean   2007.792704  
std       1.317834  
min    2006.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2010.000000  

[8 rows x 38 columns]

In [10]:
# Summary statistics (categorical columns)
housing_df.describe(include="object")

MS_Zoning Street Alley Lot_Shape Land_Contour Utilities Lot_Config  \
count       2933   2933   198      2933         2933      2933       2933   
unique         7      2     2         4            4         3          5   
top           RL   Pave  Grvl       Reg          Lvl    AllPub     Inside   
freq        2276   2921   120      1861         2636      2930       2143   

       Land_Slope Neighborhood Condition_1  ... Garage_Type Garage_Finish  \
count        2933         2933        2933  ...        2776          2774   
unique          3           28           9  ...           6             3   
top           Gtl        NAmes        Norm  ...      Attchd           Unf   
freq         2792          446        2524  ...        1734          1233   

       Garage_Qual Garage_Cond Paved_Drive Pool_QC  Fence Misc_Feature  \
count         2774        2774        2933      13    573          106   
unique           5           5           3       4      4            5   
top             TA          TA           Y      Ex  MnPrv         Shed   
freq          2618        2668        2655       4    330           95   

       Sale_Type Sale_Condition  
count       2933           2933  
unique        10              6  
top          WD          Normal  
freq        2539           2416  

[4 rows x 43 columns]

From the above summary statistics we can derive lots of useful information from this dataset (both numerical and categorical columns) such as:
- The maximum Order number is 2930 while we have 2933 rows, check for duplicates
- Lot area has a maximum value of 215,245 while the average is 10,000. Check for outliers 
- The oldest house in the dataset was built in 1872, it would be interesting to see how much this was sold for
- The largest property has 8 bedrooms, identify which property this is and possibly its sale price compared to rest
- 
- Within the categorical columns, we can see the number of unique values each column had as well as how many times the most frequent value appeared
    - This will be useful in identifying the most popular values within the dataset
    - We can also see the number of missing values within each column

In [11]:
# Summary statistics
sales_df.describe()

Order           PID  MS_SubClass      SalePrice
count  2933.000000  2.933000e+03  2933.000000    2933.000000
mean   1464.151381  7.142811e+08    57.349131  180759.629731
std     846.579702  1.887213e+08    42.632965   79858.300395
min       1.000000  5.263011e+08    20.000000   12789.000000
25%     731.000000  5.284770e+08    20.000000  129500.000000
50%    1464.000000  5.354532e+08    50.000000  160000.000000
75%    2197.000000  9.071811e+08    70.000000  213500.000000
max    2930.000000  1.007100e+09   190.000000  755000.000000

Similar to the Housing_df, we can draw similar insights for Order, PID and MS_Subclass but we will confirm if all the output in both dataframes are identical later on.
Based of the sales_df summary statistics we can see that:
- There is also a discrepancy between the Order numbers and maximum value
- The average sale price is 180,000 while the max is 755,000 -> This could be an outlier, do more research

### Data Dictionary

The data dictionary is a useful tool to help us understand the data and the meaning of each column. This will help us understand the data better and help us identify any potential issues with the data.

In [12]:
# Identifying all object columns in the housing_df
for name, label in housing_df.items():
    if pd.api.types.is_object_dtype(label):
        print(name)

MS_Zoning
Street
Alley
Lot_Shape
Land_Contour
Utilities
Lot_Config
Land_Slope
Neighborhood
Condition_1
Condition_2
Bldg_Type
House_Style
Roof_Style
Roof_Matl
Exterior_1st
Exterior_2nd
Mas_Vnr_Type
Exter_Qual
Exter_Cond
Foundation
Bsmt_Qual
Bsmt_Cond
Bsmt_Exposure
BsmtFin_Type_1
BsmtFin_Type_2
Heating
Heating_QC
Central_Air
Electrical
Kitchen_Qual
Functional
Fireplace_Qu
Garage_Type
Garage_Finish
Garage_Qual
Garage_Cond
Paved_Drive
Pool_QC
Fence
Misc_Feature
Sale_Type
Sale_Condition


From the above code, we can see that there are no object type columns in the dataset. From our analysis we'll need to change the following column to object type:
- MS_SubClass

In [13]:
# identifying all numeric columns in the housing_df
for name, label in housing_df.items():
    if pd.api.types.is_numeric_dtype(label):
        print(name)

Order
PID
MS_SubClass
Lot_Frontage
Lot_Area
Overall_Qual
Overall_Cond
Year_Built
Year_Remod/Add
Mas_Vnr_Area
BsmtFin_SF_1
BsmtFin_SF_2
Bsmt_Unf_SF
Total_Bsmt_SF
1st_Flr_SF
2nd_Flr_SF
Low_Qual_Fin_SF
Gr_Liv_Area
Bsmt_Full_Bath
Bsmt_Half_Bath
Full_Bath
Half_Bath
Bedroom_AbvGr
Kitchen_AbvGr
TotRms_AbvGrd
Fireplaces
Garage_Yr_Blt
Garage_Cars
Garage_Area
Wood_Deck_SF
Open_Porch_SF
Enclosed_Porch
3Ssn_Porch
Screen_Porch
Pool_Area
Misc_Val
Mo_Sold
Yr_Sold


In [14]:
# Identifying all object columns in the sales_df
for name, label in sales_df.items():
    if pd.api.types.is_object_dtype(label):
        print(name)

From the above code, we can see that there are no object(categorical) type columns in the dataset. From our analysis we'll need to change the following column to object type:

In [15]:
# identifying all numeric columns in the sales_df
for name, label in sales_df.items():
    if pd.api.types.is_numeric_dtype(label):
        print(name)

Order
PID
MS_SubClass
SalePrice


Based of the code above, we can make the following conclusions about the variables we have in our dataset


|  **Variable**  |  **Description**  | **Type** | **Subtype** | 
|:--------------:|:-----------------:|:--------:|:-----------:|
| Order | Observation Number | Numerical |  Discrete   |
| PID | Parcel Identification Number | Numerical |   Nominal   |
| MS_SubClass | Identifies the type of dwelling involved in the sale | Numerical |   Nominal   |
| MS_Zoning | Identifies the general zoning classification of the sale | Categorical |   Nominal   |
| Lot_Frontage | Linear feet of street connected to property | Numerical | Continuous  |
| Lot_Area | Lot size in square feet | Numerical | Continuous  |
| Street | Type of road access to property | Categorical |   Nominal   |




## 3. Data Preparation

This stage will include:
- Initial Data cleaning (checking for duplicates)
- Merging the 2 datasets
- Data cleaning
    - Checking for missing values, outliers, duplicates
- Data transformation
    - Adjusting data types
    - Creating new columns from existing ones
    - Dropping columns that aren't crucial

### 3.1 Initial Data Cleaning

This stage will check for duplicates in both dataframes and remove them if they exist before merging the 2 datasets.

In [18]:
# Housing_df duplicates
housing_df[housing_df.duplicated(keep=False)].sort_values(by="Order")

Order        PID  MS_SubClass MS_Zoning  Lot_Frontage  Lot_Area Street  \
145    146  535175070           20        RL          73.0      9300   Pave   
148    146  535175070           20        RL          73.0      9300   Pave   
146    147  535175180           20        RL          87.0     10725   Pave   
149    147  535175180           20        RL          87.0     10725   Pave   
147    148  535179020           20        RL          80.0     10032   Pave   
150    148  535179020           20        RL          80.0     10032   Pave   

    Alley Lot_Shape Land_Contour  ... Screen_Porch Pool_Area Pool_QC Fence  \
145   NaN       Reg          Lvl  ...          143         0     NaN   NaN   
148   NaN       Reg          Lvl  ...          143         0     NaN   NaN   
146   NaN       IR1          Lvl  ...            0         0     NaN   NaN   
149   NaN       IR1          Lvl  ...            0         0     NaN   NaN   
147   NaN       Reg          Lvl  ...          160         0     NaN  GdWo   
150   NaN       Reg          Lvl  ...          160         0     NaN  GdWo   

    Misc_Feature Misc_Val Mo_Sold Yr_Sold  Sale_Type  Sale_Condition  
145          NaN        0       4    2010        WD           Normal  
148          NaN        0       4    2010        WD           Normal  
146          NaN        0       5    2010        WD           Normal  
149          NaN        0       5    2010        WD           Normal  
147          NaN        0       6    2010        WD           Normal  
150          NaN        0       6    2010        WD           Normal  

[6 rows x 81 columns]

From the above code, we can see that there are 3 duplicates in the housing_df dataset. We will remove these duplicates before merging the 2 datasets.

In [19]:
# Dropping duplicates
housing_df.drop_duplicates(inplace=True)

# Checking if we have any duplicates
housing_df.duplicated().sum()

0

In [20]:
# Sales_df duplicates
sales_df[sales_df.duplicated(keep=False)].sort_values(by="Order")

Order        PID  MS_SubClass  SalePrice
145    146  535175070           20     167500
148    146  535175070           20     167500
146    147  535175180           20     108538
149    147  535175180           20     108538
147    148  535179020           20     159500
150    148  535179020           20     159500

In [21]:
# Dropping duplicates
sales_df.drop_duplicates(inplace=True)

# Checking if we have any duplicates
sales_df.duplicated().sum()

0

Similarly,we can see that there were 3 duplicates in the sales_df dataset. We will remove these duplicates before merging the 2 datasets.

### 3.2 Merging the 2 datasets

Before merging the dataframes, we need to ensure that we have matching PID and MS SubClass in both dataframes

In [22]:
# Comparing whether the PID values and MS Subclass are identical in both df

equal_pid = housing_df["PID"].isin(sales_df["PID"]).all()
equal_subclass = housing_df["MS_SubClass"].isin(sales_df["MS_SubClass"]).all()

if equal_pid and equal_subclass:
    print("Both PID and MS SubClass are identical in both df")
else:
    print("They aren't identical. Perform further analysis")

Both PID and MS SubClass are identical in both df


In [23]:
# Merging both the housing and sales dataframes
df = housing_df.merge(sales_df, on=["PID","MS_SubClass","Order"],
                      how="inner")
df.head()

Order        PID  MS_SubClass MS_Zoning  Lot_Frontage  Lot_Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot_Shape Land_Contour  ... Pool_Area Pool_QC  Fence Misc_Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc_Val Mo_Sold Yr_Sold Sale_Type  Sale_Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [24]:
# Checking the length of the dataframes
print(f"Length of the housing_df is: {len(housing_df)}")
print(f"Length of the sales_df is: {len(sales_df)}")
print(f"Length of the merged df is: {len(df)}")

Length of the housing_df is: 2930
Length of the sales_df is: 2930
Length of the merged df is: 2930


### 3.3 Checking for missing values

This stage will check for missing values in the merged dataframe and deal with them accordingly.

In [26]:
# Total number of NA values
print("Total number of missing values in the dataframe:")
df.isna().sum().sum()

Total number of missing values in the dataframe:


15749

We will deal with the missing values in different ways depending on the column:
1. Either drop the NA values
2. Impute the missing values
3. Do nothing to them if they are valid values

In [27]:
# Columns with na values
nan_columns = df.columns[df.isna().any()].tolist()

# Sum of the na values in the dataset
nan_values = df[nan_columns].isna().sum()

# Percentage of na values
nan_percentage = (nan_values / len(df)) * 100

# Creating a dataframe to display the results
nan_df = pd.DataFrame({"Missing values": nan_values,
                       "Percentage of missing values": nan_percentage})

# Sorting the dataframe
nan_df.sort_values(by="Percentage of missing values", ascending=False, inplace=True)
nan_df

Missing values  Percentage of missing values
Pool_QC                   2917                     99.556314
Misc_Feature              2824                     96.382253
Alley                     2732                     93.242321
Fence                     2358                     80.477816
Mas_Vnr_Type              1775                     60.580205
Fireplace_Qu              1422                     48.532423
Lot_Frontage               490                     16.723549
Garage_Cond                159                      5.426621
Garage_Qual                159                      5.426621
Garage_Finish              159                      5.426621
Garage_Yr_Blt              159                      5.426621
Garage_Type                157                      5.358362
Bsmt_Exposure               83                      2.832765
BsmtFin_Type_2              81                      2.764505
Bsmt_Cond                   80                      2.730375
Bsmt_Qual                   80                      2.730375
BsmtFin_Type_1              80                      2.730375
Mas_Vnr_Area                23                      0.784983
Bsmt_Half_Bath               2                      0.068259
Bsmt_Full_Bath               2                      0.068259
BsmtFin_SF_1                 1                      0.034130
Garage_Cars                  1                      0.034130
Garage_Area                  1                      0.034130
Total_Bsmt_SF                1                      0.034130
Bsmt_Unf_SF                  1                      0.034130
BsmtFin_SF_2                 1                      0.034130
Electrical                   1                      0.034130

Looking at the above dataframe, we can see the amount of NA values in the dataset as well as its proportion to the whole dataset. 
For this project, we will drop all columns with **more than 10% of missing values**. This is because we have a large dataset and imputing over them will cause a lot of noise in the data.

In addition, dropping them makes sense over imputing them as some columns such as `Pool_QC` and `Misc_Feature` have more than 90% of missing values. Looking at the data dictionary we see that "NA" means that the property doesn't have the feature or pool. Therefore, we can safely assume that the missing values are valid and we can drop them.

#### 3.3.1 Dropping columns with more than 10% of NA values


In [29]:
# Dropping columns with more than 10% of NA values
# Columns to drop
columns_to_drop = nan_df[nan_df["Percentage of missing values"] > 10].index.tolist()
print(columns_to_drop)

# Drop columns
housing_df.drop(columns=columns_to_drop, inplace=True)

print(f"Number of columns in the dataset was: {len(housing_df.columns) + len(columns_to_drop)}")
print(f"Number of columns to drop is: {len(columns_to_drop)}")
print(f"Number of columns in the updated dataset is: {len(housing_df.columns)}")

['Pool_QC', 'Misc_Feature', 'Alley', 'Fence', 'Mas_Vnr_Type', 'Fireplace_Qu', 'Lot_Frontage']
Number of columns in the dataset was: 81
Number of columns to drop is: 7
Number of columns in the updated dataset is: 74


In [30]:
# Update nan_df 
nan_df = nan_df[~nan_df.index.isin(columns_to_drop)]

# Updating the original df
df = df.drop(columns=columns_to_drop)

# Checking the shape of the df
df.shape

(2930, 75)

Looking at the shape of the original df, we can see that the number of columns has dropped from 81 to 75. This is because we dropped the columns that had more than 10% of missing values.

#### 3.3.2 Columns to leave as they are

he first set of missing values we will deal with are columns that are correctly labelled as NA.
How we identified columns labeled as NA, is by going through the data dictionary provided and seeing if any columns were meant to have an NA value (eg. If property doesn't have a pool, the value associated with it in the dictionary is NA).

In [31]:
# Correctly labelled NA values
correct_na_cols = ["Bsmt_Qual","Bsmt_Cond","Bsmt_Exposure","BsmtFin_Type_1",
            "BsmtFin_Type_2","Garage_Type","Garage_Finish","Garage_Qual",
            "Garage_Cond"]

# Remove the columns from the nan_df
nan_df = nan_df[~nan_df.index.isin(correct_na_cols)]

# View the updated df
nan_df


Missing values  Percentage of missing values
Garage_Yr_Blt              159                      5.426621
Mas_Vnr_Area                23                      0.784983
Bsmt_Half_Bath               2                      0.068259
Bsmt_Full_Bath               2                      0.068259
BsmtFin_SF_1                 1                      0.034130
Garage_Cars                  1                      0.034130
Garage_Area                  1                      0.034130
Total_Bsmt_SF                1                      0.034130
Bsmt_Unf_SF                  1                      0.034130
BsmtFin_SF_2                 1                      0.034130
Electrical                   1                      0.034130

Based of the updated dataframe, we can see that the columns above consist of values that can now be imputed with the (assumed) correct values.
The 2 main imputation methods we will use will be either:
1. Mean
2. Median

#### 3.3.3 Imputing missing values

Kumar (2023) gives us a guide on which method to use for imputing missing values depending on the type of data we have.
- **Mean Imputation**  is used for continuous data that is normally distributed
- **Median Imputation** is used for continuous data that is skewed as it is less sensitive to outliers
- **Mode Imputation** is used for categorical data
